hooks/useMqlQuery.ts

```diff
@@ -0,0 +1,305 @@
+  // We cap the list at 10. It's only intended to handle scenarios where the user makes a series of requests while a series of requests is
+  // currently in flight
+  cancelledQueries: string[];
+
+  // The time we started a new timeSeries query.
+  fetchStartTime: number | null;
+
+  // We will track how long SUCCESSFUL response is taking, if it takes too long we will set this to true.
+  isTakingForever: boolean;
+
+  errorMessage?: string;
+};
+
+const initialState: State = {
+  queryId: null,
+  queryStatus: MqlQueryStatus.Pending,
+  chartData: null,
+  cancelledQueries: [],
+  fetchStartTime: null,
+  isTakingForever: false,
+};
+
+type Action =
+  | { type: "postQueryStart" }
+  | { type: "postQueryFail"; errorMessage: string }
+  | { type: "postQuerySuccess"; queryId: string }
+  | { type: "fetchResultsFail"; errorMessage: string }
+  | { type: "fetchResultsRunning" }
+  | {
+      type: "fetchResultsSuccess";
+      data: FetchMqlTimeSeriesQuery;
+      limit?: number;
+      handleCombinedError: (e: CombinedError) => void;
+    };
+
+function mqlQueryReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case "postQueryStart": {
+      // This condition is triggered when there is already a queryId being fetched.
+      // In this case we want to add "cancel" that query by adding it to the cancelledQueries list
+      if (state.queryId) {
+        return {
+          ...state,
+          queryStatus: MqlQueryStatus.Running,
+          cancelledQueries: [
+            state.queryId,
+            ...state.cancelledQueries.slice(CANCELLED_QUERY_LIST_LENGTH - 1),
+          ],
+          chartData: null,
+          fetchStartTime: Date.now(),
+          isTakingForever: false,
+          errorMessage: undefined,
+        };
+      }
+      return {
+        ...state,
+        queryStatus: MqlQueryStatus.Running,
+        fetchStartTime: Date.now(),
+        isTakingForever: false,
+        chartData: null,
+        errorMessage: undefined,
+      };
+    }
+
+    case "postQueryFail": {
+      return {
+        ...state,
+        queryStatus: MqlQueryStatus.Failed,
+        errorMessage: action.errorMessage,
+      };
+    }
+
+    case "postQuerySuccess": {
+      return {
+        ...state,
+        queryId: action.queryId,
+        errorMessage: undefined,
+      };
+    }
+
+    case "fetchResultsRunning": {
+      const now = Date.now();
+      const diff = now - (state.fetchStartTime || 0);
+      return {
+        ...state,
+        isTakingForever: diff >= LONG_FETCH_QUERY_ATTEMPT_MAX,
+        errorMessage: undefined,
+      };
+    }
+
+    case "fetchResultsFail": {
+      return {
+        ...state,
+        queryStatus: MqlQueryStatus.Failed,
+        isTakingForever: false,
+        errorMessage: action.errorMessage,
+      };
+    }
+
+    case "fetchResultsSuccess": {
+      let chartData: Serie[];
+      try {
+        chartData = formatChartData(action.data.mqlQuery, action.limit);
```

Comment:
I wonder if we shouldn't include this part?  The formatChartData func assumes the chart data will be put into a nivo chart and adds colors to the data objects.  I wonder if this would be confusing for our end users?  Perhaps we should just return the data as it is returned from the mql server?

Reply:
Right. Seems reasonable.

Reply:
@jjrajani I have moved the chart formatting and serie concept out of `useMqlQuery`. 

I think we probably want to include a chart in the example, so perhaps it would be good to leave those in place and fast follow with a PR that uses Nivo and shows how to format the data for chart rendering. WDYT?

In either case, This is definitely displaying data, so if you think it's good to go, I can merge this version as-is and keep iterating!

Reply:
Ah, I didn't see this before leaving my comments about not including formatChartData or colors in the package...

I guess I don't have a problem keeping those pieces in here if they aren't exposed through the API at all.  As long as it's for demo purposes only I don't see any issue.